### DISCUSSION TITLE
Intent triggered update

### DISCUSSION THREAD

**Discussion by pingu8007** [breezy-weather/breezy-weather#1266] [Ideas]
### Describe your idea

I've read that all features like update on app start won't be accepted. Can I trigger update with intent?

I use Tasker (powered by exact alarm) to trigger AntennaPod to update podcast every 7 a.m., when is the time new episode released. 

I also using some local weather source that having similar release cycle. If I can trigger update with intent, I can let Breezy Weather update every 3hr and trigger an extra update by the time the source updated, instead of update every 30 min to get latest report.

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.7 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [X] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by pingu8007** [comment]
This idea opened without any interaction for 5 months, bump up.

I'm not Android app developer, but since the background update and the scheduler are already there, I think a intent listener to trigger the update shouldn't be a big problem?

With the intent listener, BW may be integrated by any app. With integration to others, update on wifi connection, by exact time, by any rule become possible. BW only need to maintain the listener and, if necessary, the concurrent update lock.

**Reply by papjul** [reply]
1. Can you explain the link between the podcast and Breezy Weather? Or was it just an example of something that you can do with Tasker? Because in the next sentence, you say the weather source has a similar schedule to the podcast, so I'm not sure of the link between the two.
2. When you say "local weather source", is it Breezy Weather or a different weather app of your country/region? If it is Breezy Weather, which source are we talking about? And what is there so special with this source?
3. How is the intent at a scheduled time going to make you execute the background work less frequently?

While waiting for your answers, I already see two things:
1. It's usually not a good idea to use exact alarm. The advantage of periodic work is that Android will not execute it precisely every -let's say- 1.5 hours. It will have some margin to execute it when it feels the device is not under heavy load. But at the same time, you seem to say that this will allow you to execute it less frequently, so let's wait for your answer for this ;)
2. There would need some guardrails. Intent must not be an excuse for someone to execute it every minute when it isn't allowed in the app itself. So it cannot be executed more frequently by an intent than the minimum allowed refresh rate in the app (30 min). At the same time, it looks like we must disallow background work when intent is used. Because let's say background work and intents are both used. The background work executes at 6:48, then your intent comes at exactly 7:00, but it already ran at 6:48, so it will not be executed, leading to bug reports or you not being happy because you wanted it to be executed precisely at 7:00 for a reason.

It seems technically feasible, but it has its caveats and it's a non-trivial added logic to the app, and work to do, so I'm looking for something that is not an edge-case ;)

**Comment by pingu8007** [comment]
1. No, there is no relation. It is a example that how I use Tasker to trigger something on exact time.
2. By local weather source I mean CWA (Taiwan), which is already integrated with BW. It publish full forecast in fixed time because it's a government department.
3. Just like you get up at 5:30 everyday. There are some "peak" when people need the most fresh forecast. I guess you also have experience that the forecast is not fresh enough and update it manually.  
If a forecast has fixed or predictable update time, it's more efficiency to follow the schedule instead of pulling blindly.

To your answer
1. You're right, that's the idea behind doze, but sometime we just can't wait.
Doze is best for service without fixed update time or is rolling updated which always give latest data by the time. To service with fixed update time, since dozed schedule not always executes on the right time, the longer the upstream update duration is, the worser freshness we get.
Things getting even worse on low end devices which is more likely been killed due to battery optimization or low memory. (I have a tablet stopped on 1/15)
2. It's reasonable. At least I'm not interesting to kill my battery.
3. (New) It can also be used as a workaround for some extreme battery optimization. Comparing to BW, Tasker use foreground service and is harder to be killed.

**Reply by papjul** [reply]
*(redacted other notes that got copied by mistake in the middle, sorry)*

A few other questions:
1. What are these fixed time? Could you send the link to the documentation?
2. Is it published exactly at -let's say- 7:00? Or is there a delay like "I have the data ready at 7:00 but then I need to process it, so it won't be ready until at least 7:10 on the API"?

I'm not in favor of asking users to install a third-party app to refresh at a specific time their weather data. This brings unnecessary workload to the CWA server at a specific time and everyone is just going to get server errors. The current approach of being at a random time is the good one.

So I have "another way around" idea: what I can do instead is to make Breezy Weather ignore CWA-sourced data refresh if the last successful refresh is still the same data (based on the documented fixed time you will send me).

This way, let's say it's published at 4:00 and 7:00. And you set up refresh rate to (approximately) every 30 min.

| Time | Status |
|---|---|
| 4:12 | Refreshed |
| 4:44 | Skipped |
| 5:08 | Skipped |
| 5:40 | Skipped |
| 6:11 | Skipped |
| 6:42 | Skipped |
| 7:13 | Refreshed |

3. However, it is valid only for the forecast part, right? What about the current observations? This kind of data is refreshed every X minutes server side. Did you turn it off?
4. More generally, CWA supports Forecast, Current, Air quality, Alert and Normals. What are the fixed refresh time for all of them? I don't believe they are all the same. Even if you don't use these features, it needs to be set up correctly. We don't want someone to miss an alert because it wasn't refreshed for 3 hours

> Comparing to BW, Tasker use foreground service and is harder to be killed.

We do use a foreground service (it's mandatory for this kind of work with recent Android SDK). https://github.com/breezy-weather/breezy-weather/blob/v5.4.1-beta/app/src/main/java/org/breezyweather/background/weather/WeatherUpdateJob.kt#L107

**Comment by chunshek** [comment]
Here are the update schedules for CWA data:
* Forecast: Regular updates at 05:30, 11:30, 17:30, 23:30 (UTC+8) / Ad-hoc updates as needed
* Current: Every 10 minutes
* Air Quality: Every hour (in association with MOENV)
* Alerts: As needed
* Normals: Every 10 years

**Reply by papjul** [reply]
Thanks!

The minimum refresh rate is 30 min.
This seems to be appropriate for current, air quality and alerts.
Normals are refreshed once a month (unless you use current location, in that case we need to check your new location and normals), which is also appropriate IMO.
This leaves us only with forecast. While it seems to be refresh once every 6 hours, you mention "ad-hoc updates", does it mean there may still be relevant updates between 05:30 and 11:30? Do you know if fixed time updates are reflected immediately in the API?

**Reply by pingu8007** [reply]
~~Ad-hoc update are published once they are ready. I'm not sure if they are reflected immediately in the API, but there are no noticable intentional delay in the mail notification.~~
Miss understood. What I mentioned here is alert, not forecast.

I'm not sure if this direction (reduce unnecessary update) post any issue, as my thought was add an additional update.

**Reply by papjul** [reply]
I took this direction, because you said:
> If I can trigger update with intent, I can let Breezy Weather update every 3hr and trigger an extra update by the time the source updated, instead of update every 30 min to get latest report.

And I explained that having intent as an additional update would cause this issue:
> The background work executes at 6:48, then your intent comes at exactly 7:00, but it already ran at 6:48, so it will not be executed, leading to bug reports or you not being happy because you wanted it to be executed precisely at 7:00 for a reason.

And you would have to wait for 9:48 for an update, instead of 7:00 like you wanted, or 7:18 if you were to keep refresh rate of 30 min.

**Reply by pingu8007** [reply]
Umm...I have a bad feel but I can't tell it clearly. 
1. In case the main location is a combination of multiple sources (CWA doesn't support all data type), short update meant those generic source (e.g. AccuWeather) are still getting update, right? To get the fresh data from CWA, it cause AccuWeather receive extra request and the battery still been killed.
2. If CWA change their schedule, will BW App need to be updated?

**Reply by papjul** [reply]
1. If I follow my suggestion, and if you set your refresh rate to 30 min, you would get update for every kind of data except forecast if it is still up-to-date according to the fixed schedule. However, this means we won't receive ad-hoc forecast updates, so I'm not sure if it's a good thing (and if the original reason you wanted it is still valid given those ad-hoc updates? + given all other kind of data of CWA still need more regular updates than every 3 hours)
2. Yes

**Reply by pingu8007** [reply]
I have those goals in the mind
1. Keep overall refresh rate low to prolong the battery life
2. Keep forecast fresh on the given time (there may be multi time)
3. As less hardcode as possible

Adhoc update may be important to others, I don't want to break this.
By add extra updates, nothing is changed to user who stick with internal scheduling. They still can have refresh rate as low as 30 min without been affected by rate limit.
Intent is advanced feature and require reading some documents. Make a rate limit statement in document is acceptable to me.

I feel sorry but I disagree with the "reduce unnecessary updated" direction. It's over-customized, inconsistent to other sources, hard to bare adhoc "exceptions".

**Reply by papjul** [reply]
> I feel sorry but I disagree with the "reduce unnecessary updated" direction. It's over-customized, inconsistent to other sources, hard to bare adhoc "exceptions".

There is almost no customization with my solution, because we already have a [per-feature caching mechanism](https://github.com/breezy-weather/breezy-weather/blob/main/docs/UPDATES.md#caching).

If I suggested this solution, it's because it would be rather easy to add another layer on top of the existing system, but yes, given ad-hoc updates, we have to exclude this solution.

Instead, adding an intent system would need a significant work to refactor the code and to include guardrails.

> 1. Keep overall refresh rate low to prolong the battery life

I'm not aware of any battery life issue with the background update work. The background work is optimized and executes in a blink of a time. There should be no significant difference between someone who turn off background work and someone who enable it every 30 minutes. I have background work enabled every 1.5 hours with many sources in use, and it always says it uses *less* than 1% battery (so it can be 0.01% like 0.99%, but in the end it's non-significant), including the screen time of the app. If you have a battery life issue and you're not using the live wallpaper (we know there is a big issue with this feature), please report it. It needs to be investigated.

> 2. Keep forecast fresh on the given time (there may be multi time)

> By add extra updates, nothing is changed to user who stick with internal scheduling. They still can have refresh rate as low as 30 min without been affected by rate limit.
> Intent is advanced feature and require reading some documents. Make a rate limit statement in document is acceptable to me.

So, it should be documented that intent will not work if the background work was executed less than 30 minutes ago, is that correct?

So, if you keep your refresh rate set to 30 minutes, it should never work, right?

I think I understood you initially planned to lower the refresh rate to 3 hours, but you wanted other features to keep updating every 30 min, so you're going to keep your refresh rate as 30 min?

How would you personally plan to use it with your use case?

**Reply by pingu8007** [reply]
> There is almost no customization with my solution, because we already have a [per-feature caching mechanism](https://github.com/breezy-weather/breezy-weather/blob/main/docs/UPDATES.md#caching).
> 

Glade to hear this.

> Instead, adding an intent system would need a significant work to refactor the code and to include guardrails.

I never thought this feature need so much works.

> So, it should be documented that intent will not work if the background work was executed less than 30 minutes ago, is that correct?
> 
> So, if you keep your refresh rate set to 30 minutes, it should never work, right?

Correct, it should be documented. No, it shouldn't work if I still use the 30 min refresh rate, it will be my fault.

> I think I understood you initially planned to lower the refresh rate to 3 hours, but you wanted other features to keep updating every 30 min, so you're going to keep your refresh rate as 30 min?
> 
> How would you personally plan to use it with your use case?

The story start from "having latest forecast before I'm out for work". I already have a Tasker job to make everything ready by that time, and this job only fired once per day.

No, I'm not plan to make BW refreshed every 30 minute. If I set to 3 hours, the dozed update often fired too early (before new forecast published) or too late (needs no longer exist). Set to 30 minute give it more chance to get the updated forecast by that time -- until the new issue emerged.

I found BW often be killed on low end device (due to battery optimization or low memory, 3GB is low enough nowadays), even set refresh rate to 30 minute doesn't works (in worst case the Android UI also been killed). The exact alarm make Tasker so reliable and I start making it as a workaround to against those "optimization".

The result is much better than I expected. I begin to stop unnecessary background update, focusing on that moment.

That's the reason why I wish BW can support intent, too. The idea behind is to extract scheduling to where I can control.
